Update from original

pqarmitage opened the pull request.

pqarmitage commented:
@lzmartinico many thanks for this.

Although your issue report has been closed, I think this is a good idea, and I'd like to progress this so that it can be incorporated upstream.

I've had a quick look at the patch and noticed one or two points.
1. While you were producing your patch I updated the build system to use automake, so I'll update your patch to take account of that.
2. I think there is a potential problem with calling vrrp_print_data() from the dbus thread. The issue is that the other thread may be updating the vrrp data structures (particularly if there is a configuration reload in progress). I think the dbus thread needs to add a print_data event that can be dequeued and processed by the main thread. This would need adding semaphores around the queueing and dequeueing of events.

Also while you were producing your patch I added support for network namespaces, which I think complicates the dbus interface, since multiple instances of keepalived can now be run simultaneously, each in its own network namespace (in fact we are aware of two users who were running two instances of keepalived concurrently and the only complication with that was the names of the pid files). For running in namespaces, I append the namespace name to the syslog idents, so for an instance running in network namespace red, the syslog idents would be Keepalived_red, Keepalived_vrrp_red and Keepalived_healthcheckers_red. Would a similar idea work with Dbus, so that we would be org.keepalived_red.Vrrp1.Vrrp? What is the best way to handle two instances of keepalived running (either in the default network namespace or another namespace); is this where the Vrrp1 comes in, which could be Vrrp2 for the second keepalived instance? I think adding a global configuration parameter keepalived_instance could be useful here, and would also help with naming the pid files.

If you could give some example commands that can be used to test the Dbus functionality, that would be really helpful since I am not familiar with Dbus usage.

If you're happy this this, I'll create a Dbus branch here, and merge your patch into it, taking into account the changes for automake. I'm happy to also handle the event queueing for the vrrp_print_data() of that helps too.

lzmartinico commented:
@pqarmitage thank you, it would be great if you could add the automake changes and event queueing.

By convention DBus service names start with the name of the author and the name of the service itself. I added the 1 as a version number, as to allow an easy version bump to backward incompatible upgrades. Having multiple service names would mean creating duplicate copies of the interface and policy files as well. The best solution would be to add an instance level to the object path ( _/org/keepalived/Vrrp1/**red**/Vrrp_ rather than _org/keepalived/Vrrp1/Vrrp). If I am not mistaken the two keepalived instances should be able to own the bus name at the same time, so if you were to add a keepalived_instance variable it would be trivial to modify the object path. But I'll have to test that once I get the multiple instances running. 
Also, for clarity's sake, in my code I refer to instance as each vrrp object. What name would you use instead?

To test the DBus service you would run a command like: dbus-send --system --dest=org.keepalived.Vrrp1 --print-reply  _object_ _interface_._method_ _type_:_argument_
e.g.
 `dbus-send --system --dest=org.keepalived.Vrrp1 --print-reply /org/keepalived/Vrrp1/Vrrp org.keepalived.Vrrp1.Vrrp.PrintData`
or
 `dbus-send --system --dest=org.keepalived.Vrrp1 --print-reply /org/keepalived/Vrrp1/Instance/dp0s3/1 org.freedesktop.DBus.Properties.Get string:'org.keepalived.Vrrp1.Instance' string:'State'`
The xml interface files can be used as documentation for available methods or properties (or calling method _org.freedesktop.Introspectable.Introspect_ on the object

pqarmitage commented:
@lzmarinico I've (manually) merged your patch onto branch Dbus, and fixed one or two type issues, so that the new code now compiles without type mismatch warnings, and also made one or two style changes to make it consistent with keepalived's style.

I'll have a look at the event queueing aspect later, but in the mean time this should be enough to make any further changes to the dbus code.

For setting up different instances of keepalived, running in different namespaces, see keepalived/core/namespaces.c, which describes how to set up an environment for running multiple keepaliveds, in separate namespaces. I'm happy with your idea of _/org/keepalived/Vrrp1/red/Vrrp_ for handling namespaces. Could keepalived itself crate the duplicate copies of the interface and policy files when it starts up (and probably remove them when it shuts down)?

You mention changing the object path if running multiple keepaliveds in the same namespace; would that mean adding another element to the name after, for example, _red_ ?

I think using vrrp instance is fine, since it is referring to a single vrrp_instance configuration entry. Essentially each vrrp instance is a virtual router.

lzmartinico commented:
@pqarmitage thank you for your changes!
If we keep org.keepalived.Vrrp1 as the only dbus service name, then there is no need to duplicate the interface and policy files, that's why I suggested just changing the object path. 

I am not sure what you mean by having multiple keepaliveds in the same namespace. My understanding was that you would be running every keepalived program instance in a single namespace, so that if you have namespace, interface name and group number you can identify each unique vrrp instance. Is that incorrect?

pqarmitage commented:
Without considering namespaces, I want to run two instances of keepalived, with the following configurations:
_router1.cfg_

```
global_data {
    router_id R1
}
vrrp_instance VI_1 {
    interface eth0
    use_vmac
    virtual_router_id 1
    virtual_ipaddress {
        10.1.10.1/32
    }
}
```

and
_router2.cfg_

```
global_data {
    router_id R2
}
vrrp_instance VI_1 {
    interface eth0
    use_vmac
    virtual_router_id 2
    virtual_ipaddress {
        10.1.10.2/32
    }
}
```

Run these as
`keepalived -D -f router1.cfg -p router1.pid -r router1.vrrp.pid -c router1.check.pid`
and
`keepalived -D -f router2.cfg -p router2.pid -r router2.vrrp.pid -c router2.check.pid`

It's not clear to me how one would make router1 print it's config, or send a garp on it's VI_1, as opposed to router2's config or VI_1 instance.

Suppose now that there are two network namespaces created, netns1 and netns2, then both the above configurations could be run in both namespaces (provided they had the relevant net_namespace statements added to the configs), making 4 keepalived instances running, 2 in each of the 2 namespaces.

pqarmitage commented:
With glib2 version 2.30.1 I am finding that I get a segfault in g_bus_own_name() where it appears to call a function at address 0x0. The backtrace is:
#0  0x0000000000000000 in ?? ()
#1  0x000000358eeb880b in g_bus_own_name () from /lib64/libgio-2.0.so.0
#2  0x000000000042739a in dbus_main (unused=<optimized out>) at vrrp_dbus.c:304
#3  0x00000037cfa07d90 in start_thread () from /lib64/libpthread.so.0
#4  0x00000036fc2f119d in clone () from /lib64/libc.so.6

With glib2 version 2.46.2 I am not getting this segfault.

Do you know the cause of this problem and whether there is a workaround?

lzmartinico commented:
Thank you for the clarification. I guess the best solution is to add a router_id level to the object path, so that we have _/org/keepalived/Vrrp1/_ namespace _/_ routerid _/Vrrp_ and _/org/keepalived/Vrrp1/_ namespace _/_ routerid _/_ instance _/_ group. When we are not using multiple namespaces, should we have a default level in the object path, or do away with it?

I've been trying to set up namespaces on a VM, but I get an unknown keyword net_namespace when passing the new conf file, and variable network_namespace is set to NULL. 

I'm having difficulties installing that specific version of glib2 on my machine, I will let you know if I can figure it out as soon as possible, but if you look in the changelog between the two versions several DBus bugs have been fixed, so that could be a simple explanation

pqarmitage commented:
An example of configuration with a network namespace is:

```
net_namespace netns1
global_defs {
    router_id a1
}
...
```

i.e. net_namespace is at the outer level, and not within global_defs.

Regarding the object path, I agree with `/org/keepalived/Vrrp1/namespace/routerid /Vrrp`, and I think omit the namespace if no net_namespace is specified in the config (the problem with a default level is someone might configure a namespace with the same name as the default). I think the same omission could be done with instance if not running multiple instances in the same namespace, so the default, no namespace since instance object path would be your original `/org/keepalived/Vrrp1/Vrrp` etc. I'll introduce a new keyword for running multiple instances, rather than using router_id (since router_id defaults to the host name), and that can also be used for naming the pid files.

I've looked through the glib2 changelog but can't find an entry that suggests a fix for the segfault I am experiencing.

pqarmitage commented:
I've tracked the segfault down to `G_LOCK (lock);` near the start of g_bus_own_name().
G_LOCK(name) expands to
`(1 ? (*g_thread_functions_for_glib_use . mutex_lock) ((g_thread_use_default_impl ? ((GMutex*)(gpointer) ((&g__lock_lock)->static_mutex.pad)) : ((__extension__ ({ typedef char _GStaticAssertCompileTimeAssertion_16[(sizeof *(&((&g__lock_lock)->runtime_mutex)) == sizeof (gpointer)) ? 1 : -1]; __sync_synchronize (); (gpointer) *(&((&g__lock_lock)->runtime_mutex)); })) ? *(&((&g__lock_lock)->runtime_mutex)) : g_static_mutex_get_mutex_impl (&((&g__lock_lock)->runtime_mutex))))) : ((void)0));`

The problem is `g_thread_functions_for_glib_use.mutex_lock` is 0x0.

I'm not sure where to go from here.

pqarmitage commented:
I've found the problem of the segfault.

I need to add 

```
g_type_init();
g_thread_init(NULL);
```

at the beginning of dbus_main()

`g_thread_init()` has been deprecated since version 2.32 of glib2, and `g_type_init()` has been deprecated since version 2.36.

I can see a configure test for testing whether g_thread_init() needs to be called, and I'll look to see if I can find a configure test for needing g_type_init(). I'll then push a commit for handle these.

Clearly glib has worked to remove the need for *_init() calls. Are you aware of how we can check whether any other such calls may be needed with older versions of glib? Currently we need to support systems back to at least Centos 6, and my installation of Centos 6.8 has glib ver 2.28.8, and a 2.6.32 kernel. I've recently added patches to support a 2.6.19 kernel!

lzmartinico commented:
@pqarmitage I'm glad you figured it out! This should be useful: https://developer.gnome.org/glib/stable/glib-Version-Information.html

pqarmitage commented:
@lzmartinico I've made a few more commits, particularly around the Interface name when using vmacs, and also removing '.' characters from a path name.

One slight problem I have noticed with the naming convention, and I'm not sure of the best way to resolve it, is that an interface, e.g. dp03s, could have two vrrp instances on it with a vrid of 1, one being IPv4 and the other IPv6. At the moment, those would both generate the same path name.

Now that I have something working here, I will implement event queuing to make it thread safe. You have indicated you will also add the appropriate DBus to support reloading and vrrp instances being added and deleted. Are you in a position to also make the changes to support namespaces and multiple copies of keepalived running in a single namespace (I'll create a configuration entry to support this shortly - I don't think we can use router_id since it currently isn't mandatory, and defaults to the hostname) as discussed above. I've seen other DBus code that writes the interface XML files at run time; will that be necessary for handling namespaces and multiple copies of keepalived?

BTW, have you been able to resolve the configuration for running keepalived in a network namespace?

pqarmitage commented:
@lzmartinico I've now added the event queueing and everything now seems to work as expected.

lzmartinico commented:
@pqarmitage yes, I have been able to set up the namespaces configuration and run your recent modifications, everything seems to be working fine. As soon as you make the multiple copies running configuration option available, I should be able to push the changes to the new directory structure.
After that I'll modify the interfaces to integrate adding and deleting instances.
For the IPv4/IPv6 issue, I could add a further level to the object path (../Instance/#interface/#group/#IPv4or6 )?

lzmartinico commented:
I also have the new DBus methods to create and destroy objects ready to go. I'm not sure where they are meant to be called though. Whenever the config file is updated, do we keep track of instances that have been added or removed?

pqarmitage commented:
The function clear_diff_vrrp() deals with removing vrrp instances on a reload that have been deleted, and it is in vrrp_complete_instance where new interfaces are created. For adding interfaces it might be necessary somewhere before the point of calling clear_diff_vrrp() to check which interfaces are new (i.e. exist in vrrp_data but not in old_vrrp_data).

I am now within a few hours of committing the changes for supporting multiple instances of keepalived within one network namespace.

pqarmitage commented:
@lzmartinico I've now merged a series of commits into this branch, the main one being the addition of the keyword `instance` at the same level as `net_namespace`.

If there is an entry `instance fred`, then pidfiles will be /var/run/keepalived/keepalived_fred.pid etc.

I do think IPv4or6 needs to be included in the object path.

lzmartinico commented:
@pqarmitage I've pushed the new object directory structures, DBus method for adding and removing instances and automated the latter. We don't actually need to modify the object over DBus, but it may still be useful to have the methods stay. I've started automating the creation of new objects, it shouldn't take long. Can I ask you when would you be thinking to get this merged upstream?

pqarmitage commented:
@lzmartnico Thanks for all the updates. I'm currently looking at some backup/master switching performance issues, and once I've finished that I'll look at your DBus patches. Barring any disasters I would hope it would be by the end of the week that I can push this upstream.

pqarmitage commented:
@lzmartinico are you able to submit a pull request from your lzmartinico:dbus branch to my pqarmitage:Dbus branch, since that is where I need to merge them for testing.

lzmartinico commented:
@pqarmitage That's great to hear! Unfortunately there does seem to be a problem with the event handling code, when reloading the configuration file, even if unmodified. In the function process_method_call, the while loop hangs indefinitely when calling any DBus method (after the reload)

pqarmitage commented:
@lzmartinico First of all apologies for not managing to do this last week.

I've pulled your commits, and made a few further changes which I've pushed onto the Dbus branch.

I've renamed and moved find_new_vrrp() to dbus_add_new_instances() in vrrp_dbus.c. I've also had to change the structure of it somewhat, since it didn't work, for example in the loop setting `new[i] = NULL` `i` was never incremented, so if a match was found it was always setting `new[0] = NULL`.

Could you please test and confirm that the changes I have made to find_new_vrrp()/dbus_add_new_instances() work for you.

There are a few outstanding problems that I have found that I don't know how to resolve.
1. When the Name and State properties are called, only the interface and VRID are passed from DBus to keepalived, and not the address family. This means that if there are both IPv4 and IPv6 instances on the same interface with the same VRID, it is the information for the first vrrp instance that is always returned. It also means that SendGarp() sends on the wrong vrrp instance.
2. A non-root user can invoke the SendGarp() method and keepalived sends the garp messages. A non-root user can also invoke PrintData() and PrintStats(). Only root should be able make keepalived 'do' anything, although it is fine for a non-root user to see the properties.
3. Using d-feet I can invoke the CreateInstance() method. I passed it parameters "VX_1", "eth23", 25, 6) and it duly created an object. If I then attempt to get the name property for that, keepalived segfaults. As far as I can see, CreateInstance and DestroyInstance should only be able to be invoked by keepalived; is there any need to have them as exposed interfaces?

Regarding point 1, I cannot see where the Name and State properties are defined, and hence cannot see how to change them so that the family is also passed to keepalived. I would really appreciate it if you could add a brief explanation of how it woks to this log. Also, where is if defined that the name should be passed as a parameter to SendGarp()?

One bit I haven't been able to see how to test are the signals. Could you let me know a way of testing those.

lzmartinico commented:
@pqarmitage thank you for merging it and modifying _find_new_vrrp_. 
I can confirm that it still works as far as creating the new object. However, the event handling issue I mentioned in my previous comment is still present: any DBus call to an existing object gets stuck after sending a SIGHUP to keepalived, regardless of whether the configuration has been changed or not.

As for the outstanding problems:
1. The DBus objects don't have a state of their own, the properties are just an interface to the vrrp_t object fields. In the function _handle_dbus_msg_, if the queue entity has its action set as _DBUS_GET_property_, the iname and vrid are used to loop among VRRP instances to find the requested values. You will need an additional check for family in that loop, and to pass the family value of the objects on which the Get method is called to the entity. Rather than adding an additional field to dbus_queue_ent_t, I would recommend for consistency to always use a GVariant args with type _a{sv}_, the dictionary mapping strings to variants. You could then check for what the keys are rather than for the type of args to assign methods.
You will find method definitions within the interface files in folder _keepalived/share/dbus-1/interfaces_ 
As you can see from there _SendGarp_ takes no input argument, as it's directly called on the instance object. What I used the name for was to identify the correct vrrp element in vrrp_data. Would it be better to use interface, vrid and family for this as well? 
1. Permissions to access the bus are regulated in the security policy, within file _keepalived/etc/dbus-1/system.d/org.keepalived.Vrrp1.conf_
   To achieve the correct behaviour, in the default context the existing allow tag must be replaced with `<allow send_interface="org.freedesktop.DBus.Properties" />`
2. I originally added the _CreateInstance_ and _DestroyInstance_ as a follow up to our discussion on reloading the thread. As it turned out it wasn't necessary to use these methods for reloading the configuration, but since I had written them I left them in, just in case. The only use case I can think of would be if a program external to keepalived wanted for some reason to temporarily disable and then re-enable the DBus instance object while leaving the VRRP instance itself running. If this seems too unlikely, then they might as well be taken out.

To test the signals, just use `sudo dbus-monitor --system`. This will show all the activity in the system bus, so as soon as you trigger a state change or shut down keepalived you should see a new signal entry.

I hope all of that is clear. If you have any question @coldoch will be taking the DBus integration work over from me. It was a pleasure working with you on this.

pqarmitage commented:
@lzmartinico - many thanks for your contribution. It would be helpful if either you or @coldoch could submit a patch to do the following, since I don't know what needs to be done:
`You will need an additional check for family in that loop, and to pass the family value of the objects on which the Get method is called to the entity. Rather than adding an additional field to dbus_queue_ent_t, I would recommend for consistency to always use a GVariant args with type a{sv}, the dictionary mapping strings to variants. You could then check for what the keys are rather than for the type of args to assign methods.`

I can see that the _Name_ and _State_ properties are defined in _keepalived/share/dbus-1/interfaces/org.keepalived.Vrrp1.Vrrp.xml_, but how does DBus know that it should pass the _ifname_ and _vrid_ parameters for those properties, and to send the _Name_ property for the SendGarp method?

Since DBus invokes the _Name_ property prior to invoking _SendGarp_ it would probably be more straightforward to pass ifnamee, vrid and family.

For editing _keepalived/etc/dbus-1/system.d/org.keepalived.Vrrp1.conf_
the entry _<allow send_interface="org.freedesktop.DBus.Properties" />_ doesn't look quite right, since it doesn't mention keepalived. Could you please submit a patch with the relevant changes.

With regards to the SIGHUP issue, I will see if I can work out what is going on. There is also a problem sometimes if I down and up an interface, and then Dbus can no longer access information for the vrrp instance.

lzmartinico commented:
I have created a new branch https://github.com/lzmartinico/keepalived/tree/dbus-replace-ent-args 
This is roughly what I meant to do with the GVariant argument and adding the additional family check.
At the moment it's not working, it goes in segfault when calling any method besides PrintData and PrintStats. This could possibly be related to the same issue as the SIGHUP, so it could work when you fix that.
The fact that the property methods use ifname and vrid parameters and SendGarp uses iname are specific to our implementation; after we receive the method call, DBus is unaware of us using those properties to find the right VRRP instance. These parameters are passed to handle_dbus_msg by breaking up the DBus object path which first received the method: https://github.com/pqarmitage/keepalived/blob/Dbus/keepalived/vrrp/vrrp_dbus.c#L260
So to distinguish between two instances with the same interface and group but different families, we only need to pass in the last bit of the path as an argument, which I have done in the new branch.

I've noticed you've seen my new pull request to fix the permission error. The patch changes permissions on the default policy (so not root because that has its own rules), allowing users to call methods from the org.freedesktop.DBus.Properties, which includes the Get method we use to obtain properties.
I will also add two more interfaces to the default policy: org.freedesktop.DBus.Peer and org.freedesktop.DBus.Introspectable. These aren't used in any keepalived code at the moment, but are useful for debugging purposes

pqarmitage commented:
Many thanks for the update and the description above. I now understand how it works.

I have been playing around with this a bit today, and found that I could pass the vrrp_t\* as user_data, and that then gets returned for the properties, and that could be used for immediately finding the right vrrp instance. Unfortunately the user_data field isn't set for the SendGarp method.

Is using user_data a reasonable approach, or is it best avoided?

pqarmitage commented:
I've just remembered that there are complications when using threads and signals. By default any thread can receive a signal. `pthread_sigmask()` can be used to block signals from certain threads, so I'll add that in and see if that makes any difference.

pqarmitage commented:
I think the reason for the segfault is the line
`family = atoi(dirs[path_length-1]);`
which should be something like:
`family = !g_strcmp0(dirs[path_length-1], "IPv4") ? AF_INET : !g_strcmp0(dirs[path_length-1], "IPv6") ? AF_INET6 : AF_UNSPEC;`

I've made some minor mods in my code and the `Name` and `State` properties are now working.

lzmartinico commented:
I can't think of any reasons why you wouldn't want to utilise user_data, but I can't really see how you would be using it. I think the get_vrrp_instance function you wrote works fine. 
The sigmask doesn't seem to have had any effect in regard to the reloading problem.
Thanks for spotting the bug there! I will open a pull request with my GVariant argument modification for the queue entities, it may not look much cleaner but if the interface gets extended in the future it will save us some headaches.
